packages/toolkit/src/query/react/buildHooks.ts

```diff
@@ -502,6 +512,87 @@ type GenericPrefetchThunk = (
   options: PrefetchOptions
 ) => ThunkAction<void, any, any, AnyAction>
 
+/**
+ * @internal
+ */
+type CreateSuspendableQueryOptions<
+  Definitions extends EndpointDefinitions,
+  Key extends keyof Definitions
+> = {
+  name: Key
+  isSkipped: boolean
+  args: any
+  api: Api<any, Definitions, any, any, CoreModule>
+  prefetch: (arg: any, options?: PrefetchOptions | undefined) => void
+  queryStateResults: UseQueryStateResult<
+    QueryDefinitionOf<Key, Definitions>,
+    UseQueryStateDefaultResult<QueryDefinitionOf<Key, Definitions>>
+  >
+}
+
+type QueryDefinitionOf<
+  Key extends keyof Definitions,
+  Definitions extends EndpointDefinitions
+> = Definitions[Key] extends QueryDefinition<infer A, infer B, infer C, infer D>
+  ? QueryDefinition<A, B, C, D>
+  : never
+
+const createSuspendablePromise = <
+  Definitions extends EndpointDefinitions,
+  Key extends QueryKeys<Definitions>
+>({
+  isSkipped,
+  args,
+  prefetch,
+  api,
+  name,
+  queryStateResults,
+}: CreateSuspendableQueryOptions<
+  Definitions,
+  Key
+>): Suspendable['getSuspendablePromise'] => {
+  const retry = () => {
+    prefetch(args, {
+      force: true,
+    })
+  }
+
+  return (): Promise<unknown> | undefined => {
+    // We do not suspend if a query is skipped:
+    // @see https://github.com/vercel/swr/pull/357#issuecomment-627089889
+    if (!isSkipped && typeof queryStateResults.data === 'undefined') {
+      if (queryStateResults.isLoading) {
+        let pendingPromise = api.util.getRunningOperationPromise(name, args)
+
+        if (!pendingPromise) {
+          prefetch(args, {
+            force: true,
+          })
+
+          pendingPromise = api.util.getRunningOperationPromise(
+            name as any,
+            args
+          )
+
+          if (!pendingPromise) {
+            throw new Error(
+              `[rtk-query][react]: invalid state error, expected getRunningOperationPromise(${name}, ${queryStateResults.requestId}) to be defined`
+            )
+          }
```

Comment:
Do you see this ever happening? Also, are we saving another promise than `prefetch` returns into the pending promises? Otherwise we could reuse that.

Reply:
> Do you see this ever happening?

No , it's an invariant that was useful while iterating on this feature. I can remove it.

> Also, are we saving another promise than prefetch returns into the pending promises? Otherwise we could reuse that.

usePrefetch() returns a consumer:

```ts
const prefetch: (arg: any, options?: PrefetchOptions | undefined) => void
```

I do not think that I can extract a promise from that.

I've skimmed a bit the type of `api.util.prefetch` and it seems that has type:

```ts
<EndpointName extends QueryKeys<Definitions>>(
      endpointName: EndpointName,
      arg: any,
      options: PrefetchOptions
    ): ThunkAction<void, any, any, AnyAction>
 ```

How can I extract a promise from that thunk action, I've not found a test that unwraps a promise.

Can you please provide more details?

Reply:
We would need a `return dispatch`  here everywhere this has only a `dispatch` to get that promise reliably: https://github.com/reduxjs/redux-toolkit/blob/e7a98eba3fbef2137109dc68caff305f603f4d1d/packages/toolkit/src/query/core/buildThunks.ts#L453-L468 

Perfectly viable to just add that.

Reply:
It's a breaking change

Reply:
returning something from a function that returned `undefined` before? I wouldn't really call that breaking

Reply:
prefetch thunk and usePrefetch callback now return an object.

```ts
interface PrefetchActionCreatorResult {
  /**
   * Returns a promise that **always** resolves to `undefined`
   * when there are no pending requests initiated by input thunk.
   */
  unwrap(): Promise<void>,
  /**
   * Cancels pending requests.
   */
  abort(): void
}
```